backfill/2024-03-23-clients_last_seen/telemetry_derived_clients_last_seen_v2_20240322/backfill_clients_last_seen_v2.py

```diff
@@ -0,0 +1,330 @@
+  help="Issue a dry run"
+)
+parser.add_argument(
+    "--start_date",
+    "--start-date",
+    "-s",
+    help="First date to be backfilled",
+    type=click.DateTime(formats=["%Y-%m-%d"]),
+    default='2023-03-23'
+)
+parser.add_argument(
+    "--end_date",
+    "--end-date",
+    "-e",
+    help="Last date to be backfilled",
+    type=click.DateTime(formats=["%Y-%m-%d"]),
+    # default=str(date.today()),
+    default='2023-03-23',
+)
+
+
+def _backfill_staging_table(client, job_config, project_id, dataset, table, bigquery_schema, submission_date, sample_id):
+    """Backfill for a submission_date, sample_id combination."""
+    full_table_id = f"{project_id}.{dataset}.{table}_{sample_id}"
+    try:
+        table = client.get_table(full_table_id)
+    except NotFound:
+        table = bigquery.Table(full_table_id)
+        table.schema = bigquery_schema
+
+    if not table.created:
+        client.create_table(table)
+        click.echo(f"Destination table {full_table_id} created.")
+
+    print(f"Running a backfill for sample_id={sample_id} to {full_table_id}")
+    print(PARTITION_QUERY.format(submission_date=submission_date, sample_id=sample_id, full_table_id=full_table_id))
+    client.query(
+        PARTITION_QUERY.format(submission_date=submission_date, sample_id=sample_id, full_table_id=full_table_id),
+        job_config=job_config
+    ).result()
+
+def get_bigquery_schema(schema_yaml):
+    bigquery_schema = []
+    for item in schema_yaml:
+        if item['type'] == 'RECORD':
+            fields = []
+            for field in item['fields']:
+                fields.append({
+                    'mode': field['mode'],
+                    'type': field['type'],
+                    'name': field['name'],
+                })
+            bigquery_schema.append({
+                'mode': item['mode'],
+                'type': item['type'],
+                'name': item['name'],
+                'fields': fields
+            })
+        else:
+            bigquery_schema.append({
+                'mode': item['mode'],
+                'type': item['type'],
+                'name': item['name'],
+            })
+    return bigquery_schema
+
+
+def main():
+    """Backfill table `backfills_staging_derived.telemetry_derived_clients_last_seen_v2_20240322` in parallel."""
+    args = parser.parse_args()
+
+    client = bigquery.Client(args.project_id)
+
+    if args.dry_run:
+      print("Do a dry run")
+      job_config = bigquery.QueryJobConfig(dry_run=True, use_query_cache=False)
+    else:
+      job_config = bigquery.QueryJobConfig(dry_run=False, use_query_cache=False)
+
+
+    date_range = BackfillDateRange(
+        args.start_date,
+        args.end_date,
+        range_type= PartitionType.DAY,
+    )
+
+    schema_file_path = os.path.join(this_dir, "schema.yaml")
+    with open(schema_file_path, 'r') as yaml_file:
+        schema_yaml = yaml.safe_load(yaml_file)
+
+    bigquery_schema = get_bigquery_schema(schema_yaml)
```

Comment:
why does the schema need to go through this process?

Reply:
Because creating the tables via e.g. SELECT is resulting in a random order that doesn't match the query.